### PR TITLE
BFD-3050: Upgrade JDK to version 21

### PR DESCRIPTION
**Pull request by aadedejifearless**
**JIRA Ticket:**
[BFD-3050](https://jira.cms.gov/browse/BFD-3050)

**User Story or Bug Summary:**

---

### What Does This PR Do?

- Updates `apps/pom.xml`.
-  Updates `ci-java.yml` to use 21 in java-version  and references the `BASE_ANSIBLE_TAG` to use JDK 21
-  Updates `JenkinsFile`
- Updates `ops/DockerFile` to install java-21-amazon-corretto-devel`
- Updates `ops/ansible/playbooks-ccs/roles/install_java/tasks/main.yml` by fixing Error in Comment and referencing latest java-21 update.
- Updates `ops/ansible/playbooks-ccs/roles/platinum/tasks/install-dependencies.yml` to reference latest java-21 update.
- Updates `ops/ansible/playbooks-ccs/bfd-db-migrator/tasks/migrator.yml` to reference latest java-21 update.
- Updates `ops/ansible/roles/bfd-pipeline/tasks/etl.yml` to reference latest java-21 update.
- Updates `ops/ansible/roles/bfd-server/tasks/install.yml` to reference latest java-21 update.
- Updates `ops.deploy-css.groovy` to reference latest java-21 update.
- Updates `ops/jenkins/bfd-build-apps/Jenkinsfile` to reference latest java-21 update.
- Updates `ops/jenkins/bfd-build-deploy/Jenkinsfile` to reference latest java-21 update.
- Updates `ops/jenkins/bfd-build-deployment-process/Jenkinsfile`
- Updates `ops/jenkins/bfd-build-platinum/Jenkinsfile`
- Updates `ops/jenkins/bfd-cbc-build/build-push.sh` to reference latest java-21 update.
- Updates `ops/jenkins/bfd-cbc-build/toolchains.xml` to reference latest java-21 update.
- Updates `ops/jenkins/bfd-data-fda/Jenkinsfile` to reference latest java-21 update.
-  Updates `ops/jenkins/bfd-data-npi/Jenkinsfile` to reference latest java-21 update.
- Updates `ops/jenkins/bfd-deploy-apps/Jenkinsfile` to reference latest java-21-update.
- Updates `ops/jenkins/bfd-deploy-pipeline-terraservice/Jenkinsfile` to reference latest java-21-update.
- Updates `ops/jenkins/bfd-destroy-environment/Jenkinsfile` to reference latest java-21-update.
- Updates `ops/jenkins/bfd-jenkins-sonarqube/Jenkinsfile` to reference latest java-21-update.
- Updates `ops/jenkins/bfd-run-pipeline-oob/Jenkinsfile` to reference latest java-21-update.
- Updates `ops/packer/build_bfd-all.json` to reference latest java-21-update.
-  Updates `ops/packer/build_bfd-platinum.json` to reference latest java-21-update

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: .github/workflows/ci-java.yml [hunk 1]
```diff
@@ -123,7 +123,7 @@ jobs:
           SHORT_SHA="$(git rev-parse --short HEAD)"
           FULL_BRANCH_TAG="${IMAGE_NAME}:${SANITIZED_REF}" # human readable, branch name tag
           FULL_COMMIT_TAG="${IMAGE_NAME}:${SHORT_SHA}"     # a short sha suffixed image tag
-          docker build . --tag "$FULL_BRANCH_TAG"
+          docker build . --tag "$FULL_BRANCH_TAG" --build-arg BASE_ANSIBLE_TAG="a861fa7"
```
Comment (lsmitchell) [inline .github/workflows/ci-java.yml:126]: Assuming we'll need to remove this before merge; just a note so we don't forget.

File: apps/pom.xml [hunk 2]
```diff
@@ -727,7 +734,7 @@
                     <configuration>
                         <toolchains>
                             <jdk>
-                                <!-- Use a Java 17 JDK, by default. -->
+                                <!-- Use a Java 21 JDK, by default. -->
```
Comment (lsmitchell) [inline apps/pom.xml:737]: Maybe just nix this comment so we don't have to keep updating it. The version is defined in the variable.

**Review by lsmitchell — COMMENTED** [review]
just a couple small things

**Review by brianburton — APPROVED** [review]
Looks good to me.  I built locally using JDK 21 and was able to run migrator, pipeline, and server successfully.

File: .github/workflows/ci-java.yml (hunk 1 above)
Comment (aadedejifearless) [reply .github/workflows/ci-java.yml:126]: Correct

File: apps/pom.xml (hunk 2 above)
Comment (aadedejifearless) [reply apps/pom.xml:737]: ok. We will do

File: ops/Dockerfile [hunk 3]
```diff
@@ -39,19 +39,22 @@ LABEL org.opencontainers.image.source=https://github.com/CMSgov/beneficiary-fhir
...
 # * <https://github.com/moby/moby/issues/28614>: Also some useful info.
 #
-# NOTE: Packages sepecified below should generally be kept up-to-date
+# NOTE: Packages specified below should generally be kept up-to-date
 #       with the dependencies encoded in the BFD platinum image
 #       instructions.
-RUN yum update -y --security && \
+
+RUN rpm --import https://yum.corretto.aws/corretto.key && \
+    curl -L -o /etc/yum.repos.d/corretto.repo https://yum.corretto.aws/corretto.repo && \
```
Comment (mjburling) [inline ops/Dockerfile:47]: These steps must be replicated in `ops/ansible/playbooks-ccs/roles/platinum/tasks/install-dependencies.yml`

File: .github/workflows/ci-java.yml [hunk 4]
```diff
@@ -204,4 +204,4 @@ jobs:
     needs: mvn-verify
     uses: ./.github/workflows/ci-ansible.yml
     with:
-      role: bfd-db-migrator
+      role: bfd-db-migrator
```
Comment (mjburling) [inline .github/workflows/ci-java.yml:207]: No new line at EOF? ಠ_ಠ

File: ops/ansible/playbooks-ccs/roles/install_java/tasks/main.yml
Comment (mjburling) [inline ops/ansible/playbooks-ccs/roles/install_java/tasks/main.yml:1]: I'm not sure if this role is **ever** used. Maybe we can drop it as part of the JDK21 upgrade?

File: .github/workflows/ci-java.yml (hunk 4 above)
Comment (aadedejifearless) [reply .github/workflows/ci-java.yml:207]: Added new Line

File: ops/ansible/playbooks-ccs/roles/install_java/tasks/main.yml
Comment (aadedejifearless) [reply ops/ansible/playbooks-ccs/roles/install_java/tasks/main.yml:1]: Deleted this role.

File: .github/workflows/ci-java.yml (hunk 4 above)
Comment (mjburling) [reply .github/workflows/ci-java.yml]: Looks like it persists. Take a look at[ this resource](https://thoughtbot.com/blog/no-newline-at-end-of-file#following-the-rules-in-your-editor).

Comment (mjburling) [reply .github/workflows/ci-java.yml]: I'm mistaken. Re-resolving 😉

File: .github/workflows/ci-java.yml (hunk 1 above)
Comment (brianburton) [reply .github/workflows/ci-java.yml:126]: Maybe it shouldn't be resolved until this is removed?

File: apps/pom.xml [hunk 5]
```diff
@@ -225,7 +226,7 @@
         <!-- https://github.com/GoogleContainerTools/jib/tree/master/jib-maven-plugin#quickstart -->
         <jib.version>3.4.0</jib.version>
         <!-- The base image (FROM) used to build images with jib. -->
-        <jib.from>amazoncorretto:17.0.7-alpine</jib.from>
+        <jib.from>amazoncorretto:21.0.0-alpine</jib.from>
```
Comment (brianburton) [inline apps/pom.xml:229]: An image for 21.0.1 is available now.

File: .github/workflows/ci-java.yml (hunk 1 above)
Comment (aadedejifearless) [reply .github/workflows/ci-java.yml:126]: What is the "it" you're referencing? The ticket or something else? @brianburton

File: apps/pom.xml (hunk 5 above)
Comment (aadedejifearless) [reply apps/pom.xml]: Updated

**Review by mjburling — APPROVED** [review]

**Review by brianburton — APPROVED** [review]

**Pull request merged**